### PR TITLE
Nuxt dev mode support

### PR DESCRIPTION
**Pull request by leoortizz**

### CODE REVIEW

File: tsconfig.json
Comment (leoortizz) [inline tsconfig.json:12]: this is a workaround for having Nuxt as a `devDependency`, otherwise it breaks the build

**Comment by DevJoghurt** [comment]
I have tested this pull request with my more complex nuxt3 apps and have the following comments.
The current implementation does not support monorepo setup. For this you could change the call a little bit and see where the app is located and then just let the nuxt cli do everything else:
```js  
  // get root dir of nuxt app
  const cmdParams = ['dev'];
  if(dir.split(process.cwd()).length > 1){
    cmdParams.push(dir.split(process.cwd() + '/')[1])
  }
  const host = new Promise<string>((resolve) => {
    const serve = spawn(CLI_COMMAND, cmdParams);
  ...
```

I have also noticed many URL errors in my more complex apps, the following fix seems to be the way forward:
```js
 // workaround for non-standard Nuxt URLs
 if (queryString.includes("vue&")) {
   fetchURL = fetchURL.replace(/\=$/, "").replace(/\=&/, "&");
 }
```
However, I still have massive problems for example with tailwind (@nuxtjs/tailwindcss) or with the new dev tool (@nuxt/devtools). I think the problem will also occur with many other use cases, but i have not yet been able to look deeper to find out exactly where the problem lies.

**Comment by leoortizz** [comment]
> I have tested this pull request with my more complex nuxt3 apps and have the following comments. The current implementation does not support monorepo setup. For this you could change the call a little bit and see where the app is located and then just let the nuxt cli do everything else:
> 
> ```js
>   // get root dir of nuxt app
>   const cmdParams = ['dev'];
>   if(dir.split(process.cwd()).length > 1){
>     cmdParams.push(dir.split(process.cwd() + '/')[1])
>   }
>   const host = new Promise<string>((resolve) => {
>     const serve = spawn(CLI_COMMAND, cmdParams);
>   ...
> ```
> 
> I have also noticed many URL errors in my more complex apps, the following fix seems to be the way forward:
> 
> ```js
>  // workaround for non-standard Nuxt URLs
>  if (queryString.includes("vue&")) {
>    fetchURL = fetchURL.replace(/\=$/, "").replace(/\=&/, "&");
>  }
> ```
> 
> However, I still have massive problems for example with tailwind (@nuxtjs/tailwindcss) or with the new dev tool (@nuxt/devtools). I think the problem will also occur with many other use cases, but i have not yet been able to look deeper to find out exactly where the problem lies.

> The current implementation does not support monorepo setup

Thanks for the great feedback! I'll investigate the problems you mentioned.

**Comment by leoortizz** [comment]
> I have tested this pull request with my more complex nuxt3 apps and have the following comments. The current implementation does not support monorepo setup. For this you could change the call a little bit and see where the app is located and then just let the nuxt cli do everything else:
> 
> ```js
>   // get root dir of nuxt app
>   const cmdParams = ['dev'];
>   if(dir.split(process.cwd()).length > 1){
>     cmdParams.push(dir.split(process.cwd() + '/')[1])
>   }
>   const host = new Promise<string>((resolve) => {
>     const serve = spawn(CLI_COMMAND, cmdParams);
>   ...
> ```

I tried a monorepo with Nuxt3 projects and also had issues running it. Turns out the errors were also happening by running the Nuxt project directly, without Firebase. I found the fixes for those errors here: https://github.com/nuxt/nuxt/issues/14146#issuecomment-1397344559

Does this fixes your issues with monorepos? If not can you share more details about the errors you had and ways to reproduce it?

> I have also noticed many URL errors in my more complex apps, the following fix seems to be the way forward:
> 
> ```js
>  // workaround for non-standard Nuxt URLs
>  if (queryString.includes("vue&")) {
>    fetchURL = fetchURL.replace(/\=$/, "").replace(/\=&/, "&");
>  }
> ```
> 
> However, I still have massive problems for example with tailwind (@nuxtjs/tailwindcss) or with the new dev tool (@nuxt/devtools). I think the problem will also occur with many other use cases, but i have not yet been able to look deeper to find out exactly where the problem lies.

I was able to reproduce the errors with `@nuxtjs/tailwindcss` and the small change I pushed fixed that. Can you try and see if your errors are gone? If not can you share the errors you get?

For `@nuxt/devtools` the problem seems to be related to our proxy which doesn't support WebSockets. The WebSockets support is being implemented.

**Comment by DevJoghurt** [comment]
@leoortizz 
Yes, I can confirm that even my larger projects are now working as expected.

Regarding the monorepo setup, I currently only tested yarn or npm with workspaces, there the error seems to be different.
I was able to fix the issue by not explicitly setting the CLI_COMMAND to the path of node_modules, but simply using nuxt dev directly. Is there a special reason for taking the path ("node_modules/.bin/nuxt")?

File: src/frameworks/nuxt/index.ts
Comment (jamesdaniels) [inline src/frameworks/nuxt/index.ts:42]: This should return void no?

**Review by jamesdaniels — CHANGES_REQUESTED** [review]

File: src/frameworks/nuxt/index.ts
Comment (leoortizz) [reply src/frameworks/nuxt/index.ts:42]: probably yes, I'll refactor this function as it got messy with my changes

File: src/test/frameworks/nuxt/helpers.ts
Comment (TheIronDev) [inline src/test/frameworks/nuxt/helpers.ts:1]: Nit: this file seems unnecessary if its only being used in the spec file.

File: src/frameworks/nuxt/index.ts
Comment (TheIronDev) [reply src/frameworks/nuxt/index.ts:87]: Thanks for writing tests for this!

File: src/test/frameworks/nuxt/helpers.ts
Comment (leoortizz) [reply src/test/frameworks/nuxt/helpers.ts:1]: @TheIronDev it's also being used here: https://github.com/firebase/firebase-tools/blob/a476540d2a908bfd5168e5342dfdf4c1a6c5040b/src/frameworks/nuxt/index.ts#L14

File: src/frameworks/nuxt/index.ts
Comment (leoortizz) [reply src/frameworks/nuxt/index.ts:87]: Thanks for pushing for it 🤝

**Review by TheIronDev — APPROVED** [review]
Approving. I'm Awaiting a James Approval before merging this in.

File: src/test/frameworks/nuxt/helpers.ts
Comment (leoortizz) [reply src/test/frameworks/nuxt/helpers.ts:1]: @TheIronDev sorry, just realized you're not referring to `frameworks/nuxt/interfaces.ts` but to `test/frameworks/nuxt/helpers.ts`. I expected this file to have more stuff but it wasn't necessary. Thanks, removing it now.

**Review by jamesdaniels — APPROVED** [review]
Couple nits, I dropped them in—otherwise let's go!

**Pull request merged**
